### PR TITLE
Prevent double-patching for patched deps

### PR DESCRIPTION
**Pull request by parksb**
This PR fixes an issue where patched dependencies using catalog protocol would be double-patched, creating complex chains like `typescript@patch:typescript@patch%3Atypescript@npm%253A5.8.3#...`. "The existing test had an problem where it wasn't using the catalog protocol. As a result, there was actually a double-patching problem that went undetected in the test.

The doubled-patching happened because the reduceDependency hook would recursively process already-patched dependencies. There is state tracking logic to prevent infinite recursion and skips processing dependencies that already contain resolved versions, ensuring clean single-patch resolution.

### CODE REVIEW

File: sources/__tests__/edge-cases.test.ts [hunk 1]
```diff
@@ -28,17 +28,31 @@ describe("edge cases and special scenarios", () => {
       },
     });
 
-    await workspace.yarn.add("typescript");
```
Comment (parksb) [inline sources/__tests__/edge-cases.test.ts:31]: This line simply runs `yarn add typescript` without using the catalog protocol. If it had been `yarn add typescript@catalog:`, the existing test would have failed to match this regex due to the double-patching issue.

**Review by suhdonghwi — APPROVED** [review]
